### DISCUSSION TITLE
Which has better performance in serializing custom data structures, CAF Inspect or Protobuf?

### DISCUSSION THREAD

**Discussion by phil-skillwon** [actor-framework/actor-framework#1782] [Q&A] [answered]
I am using CAF to build radar application software, but now I have encountered a problem: the custom structure nests multiple vectors for storing binary data and interacts with actors on other hosts via a 10G Ethernet, that is, implementing custom message interaction between remote actors through the network. However, during debugging, I found that the function `bool inspect(Inspector& ser, BeamDataSingleCpi& ins)` has become a performance bottleneck. This is a serialization function implemented for custom structures. Does CAF support using FlatBuffers or Protobuf for serializing custom structures? If so, could switching to Protobuf improve the performance of `inspect`?

**Comment by Neverlord** [comment]
Did you dig deeper into what is eating up the time in that `inpect` overload? If you have a deeply nested data structure, it could very well be constructing the C++ object that is slowing the application down. Nested vectors will cause multiple heap allocations per object and are also slow to serialize since the object is scattered across the heap, meaning that you have a low cache hit rate while traversing the object. One technique that you could consider here is using a custom allocator, e.g., with a monotonic buffer resource. That's a common technique to speed up deeply nested data structures like JSON objects. If each buffer individually is usually small, small buffer optimization could also do the job.

> Does CAF support using FlatBuffers or Protobuf for serializing custom structures? If so, could switching to Protobuf improve the performance of inspect?

Just to be sure, we're talking about the `caf::io` module, right? The direct answer to this is no, if you're using `publish` & `remote_actor`, you cannot change the internal format. However, the default network backend uses `caf::binary_serializer` and `caf::binary_deserializer`. You could specialize your `inspect` overload on those two types and operate directly on the buffers. Basically just passing bytes to/from the CAF serializers to your code.

However, I don't think Protobuf would solve the performance issues that you see. Of course it's always better to verify this with a micro benchmark. If you find Protobuf serializing and deserializing the same data structure faster than CAF, please let us know.

FlatBuffers might help. Not because of the network representation, but by improving cache locality and reducing heap allocations. But you could probably achieve the same with a custom allocator or maybe even by redesigning your data structure to avoid nested, heap-allocated containers in the first place.

Hope that helps. 🙂

**Reply by phil-skillwon** [reply]
Thank you very much, what you said has been very helpful to me. Through benchmark testing, I can confirm that the long execution time of the inspect template function I implemented is due to too many nested Vec<u8>, which has created a performance bottleneck. You mentioned earlier that in my scenario, the nested Vec<u8> significantly impacts the performance of the inspect function because it leads to multiple heap memory allocations and results in a low cache hit rate. What do you mean by "using a monotonic buffer resource"? Additionally, I would like to ask, if I use jemalloc or tcmalloc as the heap memory allocator, could it speed up the inspect?

**Reply by Neverlord** [reply, marked as answer]
I'm referring to the [monotonic_buffer_resource](https://en.cppreference.com/w/cpp/memory/monotonic_buffer_resource) from the standard library (have a look at what's in `std::pmr`). If you have containers that only grow (ideally immutable after construction), they are a good choice. Although you might want to switch from `vector` to `list` in that case to avoid "holes" in the buffer (from growing a `vector`) unless you can always `reserve` the proper size right away.

As for other heap allocators like `tcmalloc`: it really depends on the allocation patterns in your application. Maybe plugging in an alternative `malloc` implementation will speed up your application enough. But the only thing I can recommend here is trying it out and measuring it.

**Reply by phil-skillwon** [reply]
Fantastic answer, I think I've found the answer. The reason for the low serialization performance of my data structure is firstly due to the deep nesting of the data structure, especially for vectors, where there is multiple nesting, because I simply used vector<vector> to implement a two-dimensional array. Secondly, it's because the data size is too large, at least reaching 4MB. Based on these two points, I think I can optimize this data structure soon.

**Reply by phil-skillwon** [reply]
Sorry, it seems there was a problem with my previous debugging approach, and I think I've found the reason affecting the serialization performance. My custom struct indeed nested several layers of Vec, but what ultimately affected the performance was the underlying Vec<IqItem>, here is the definition of IqItem: 
struct IqItem 
{
    float i;
    float q;
};
When I only changed Vec<IqItem> to Vec<byte> and performed a type cast at the point where data is inserted, that is, casting Vec<IqItem>& to Vec<byte>& during data storage, and then transmitted it via a remote actor created by remote_actor, my benchmark concluded that: for the same 38400 bytes of binary data, the performance of inspect improved by at least 5 times, from 11ms to 2ms. I'm not sure if there's a problem with my implementation of the inspect function, or if CAF has an issue with serializing custom data structures to byte arrays by copying elements individually. Could you help me clarify this?

**Reply by Neverlord** [reply]
I don't think I can really follow your train of thought here. If you could provide a reproducer, I could take a look tho.
